src/kernel/include/thd.h

```diff
@@ -443,73 +394,29 @@ thd_init(void)
 	// assert(offsetof(struct thread, regs) == 4); /* see THD_REGS in entry.S */
 }
 
-static inline int
-curr_invstk_inc(struct cos_cpu_local_info *cos_info)
-{
-	return cos_info->invstk_top++;
-}
-
-static inline int
-curr_invstk_dec(struct cos_cpu_local_info *cos_info)
-{
-	return cos_info->invstk_top--;
-}
-
-static inline int
-curr_invstk_top(struct cos_cpu_local_info *cos_info)
-{
-	return cos_info->invstk_top;
-}
-
 static inline struct comp_info *
 thd_invstk_current(struct thread *curr_thd, unsigned long *ip, unsigned long *sp, struct cos_cpu_local_info *cos_info)
 {
-	/* curr_thd should be the current thread! We are using cached invstk_top. */
-	struct invstk_entry *curr;
-
-	curr = &curr_thd->invstk[curr_invstk_top(cos_info)];
-	*ip  = curr->ip;
-	*sp  = curr->sp;
-
-	return &curr->comp_info;
+	return chal_invstk_current(curr_thd, ip, sp, cos_info);
```

Comment:
If a shadow entry is added on, I'm missing why this is required. My mental model (likely missing something) is that if you make the shadow entry on the stack, the most code paths should work without modifications.